src/main/java/io/knowledgeassets/myskills/server/user/profile/UserProfileService.java

```diff
@@ -0,0 +1,240 @@
+	 * @param referenceId user reference id
+	 * @return MS Word document with anonymous user profile data
+	 */
+	public InputStream getAnonymousUserProfileDocument(String referenceId) {
+
+		final User user = getUserByReferenceId(referenceId);
+
+		try (final InputStream is = getTemplate();
+			 final XWPFDocument document = new XWPFDocument(is)) {
+			replacePlaceholders(document, user);
+			final ByteArrayOutputStream b = new ByteArrayOutputStream();
+			document.write(b);
+			return new ByteArrayInputStream(b.toByteArray());
+		}
+		catch (IOException | EmptyFileException | NotOfficeXmlFileException e) {
+			throw new UserProfileDocumentException(format("An error has occurred when building user " +
+					"profile document for a user with the reference id %s", referenceId), e);
+		}
+	}
+
+	private InputStream getTemplate() throws IOException {
+		if (StringUtils.isEmpty(this.templatePath)) {
+			return DEFAULT_TEMPLATE_RESOURCE.getInputStream();
+		}
+		else {
+			final Path path = Paths.get(this.templatePath);
+			if (path.toFile().exists()) {
+				try {
+					return new FileSystemResource(path).getInputStream();
+				}
+				catch (IOException ex) {
+					log.warn(format("The template file \"%s\" cannot be read. The default template file will be used as a fallback option.", path), ex);
+					return DEFAULT_TEMPLATE_RESOURCE.getInputStream();
+				}
+			}
+			else {
+				log.warn(format("The template file \"%s\" does not exist. The default template file will be used as a fallback option.", path));
+				return DEFAULT_TEMPLATE_RESOURCE.getInputStream();
+			}
+		}
+	}
+
+	private User getUserByReferenceId(String referenceId) {
+		return userRepository.findByReferenceId(referenceId).orElseThrow(() -> {
+			final String[] searchParamsMap = {"referenceId", referenceId};
+			return NoSuchResourceException.builder()
+					.model(Model.USER)
+					.searchParamsMap(searchParamsMap)
+					.build();
+		});
+	}
+
+	private void replacePlaceholders(XWPFDocument document, User user) {
+		replaceTablesPlaceholders(document, user);
+		replaceParagraphsPlaceholders(document, user, document.getParagraphs());
+		ofNullable(document.getFooterList()).ifPresent(footers ->
+				footers.forEach(footer -> {
+					replaceTablesPlaceholders(footer, user);
+				})
+		);
+	}
+
+	private void replaceTablesPlaceholders(IBody body, User user) {
+		ofNullable(body.getTables()).ifPresent(tables ->
+				tables.forEach(table ->
+						ofNullable(table.getRows()).ifPresent(rows ->
+								rows.forEach(row ->
+										ofNullable(row.getTableCells()).ifPresent(cells ->
+												cells.forEach((XWPFTableCell cell) ->
+														replaceParagraphsPlaceholders(cell, user, cell.getParagraphs())
+												)
+										)
+								)
+						)
+				)
+		);
+	}
+
+	private void replaceParagraphsPlaceholders(IBody body, User user, List<XWPFParagraph> paragraphs) {
+		ofNullable(paragraphs).ifPresent(paragraphList -> {
+			for (int idx = 0; idx < paragraphs.size(); idx++) {
```

Comment:
Why did you choose an index-based iteration here while you used `list.forEach()` in all the other methods? 😉

Reply:
I did this to avoid `java.util.ConcurrentModificationException`. As a new paragraph can be added into an object of `org.apache.poi.xwpf.usermodel.IBody` in the `io.knowledgeassets.myskills.server.user.profile.UserProfileDocumentService#replaceListPlaceholder` a collection of paragraphs is modified implicitly and an exception can be thrown. I decided not to use `java.util.concurrent.CopyOnWriteArrayList` and just to work with a list directly as I know that every thread has its own input stream taken from a template file. Perhaps there can be a better solution. Now I am looking for that.